File: web/src/types/graphConfig.ts
/**
 * GraphConfig TypeScript Types (v3 Canonical Schema)
 *
 * These types mirror the backend canonical schema defined in:
 * service/app/schemas/graph_config.py
 *
 * The backend backfills all agents to v3 at startup and validates on read,
 * so the frontend always receives and produces v3 configs.
 */

// =============================================================================
// Enums
// =============================================================================

/** Supported executable node kinds. Matches backend GraphNodeKind. */
export type GraphNodeKind = "llm" | "tool" | "transform" | "component";

/** Supported state field types. */
export type StateFieldType =
  | "string"
  | "int"
  | "float"
  | "bool"
  | "list"
  | "dict"
  | "any";

/** Supported state reducers. */
export type StateReducerType = "replace" | "add_messages";

/** Built-in edge conditions. */
export type BuiltinEdgeCondition = "has_tool_calls" | "no_tool_calls";

/** Supported custom predicate operators. */
export type PredicateOperator = "eq" | "neq" | "truthy" | "falsy";

// =============================================================================
// State Schema Types
// =============================================================================

export interface StateFieldSchema {
  type: StateFieldType;
  description?: string | null;
  default?: unknown;
}

export interface GraphStateConfig {
  schema?: Record<string, StateFieldSchema>;
  reducers?: Record<string, StateReducerType>;
}

// =============================================================================
// Execution Limits
// =============================================================================

export interface GraphExecutionLimits {
  max_time_s: number;
  max_steps: number;
  max_concurrency: number;
}

// =============================================================================
// Dependencies
// =============================================================================

export interface ModelDependencyRef {
  key: string;
  provider?: string | null;
  version?: string | null;
}

export interface PromptDependencyRef {
  key: string;
  version?: string | null;
}

export interface ComponentDependencyRef {
  key: string;
  version?: string;
}

export interface GraphDeps {
  models?: ModelDependencyRef[];
  tools?: string[];
  prompts?: PromptDependencyRef[];
  components?: ComponentDependencyRef[];
}

// =============================================================================
// Metadata
// =============================================================================

export interface GraphMetadata {
  display_name?: string | null;
  description?: string | null;
  tags?: string[];
  agent_version?: string | null;
}

// =============================================================================
// Node Config Types
// =============================================================================

export interface LLMNodeConfig {
  prompt_template: string;
  output_key: string;
  model_override?: string | null;
  temperature_override?: number | null;
  max_tokens?: number | null;
  tools_enabled: boolean;
  tool_filter?: string[] | null;
  max_iterations: number;
  message_key?: string | null;
}

export interface ToolNodeConfig {
  execute_all: boolean;
  tool_filter?: string[] | null;
  output_key: string;
  timeout_seconds: number;
}

export interface TransformNodeConfig {
  template: string;
  output_key: string;
  input_keys: string[];
}

export interface ComponentReference {
  key: string;
  version: string;
}

export interface ComponentNodeConfig {
  component_ref: ComponentReference;
  config_overrides: Record<string, unknown>;
}

// =============================================================================
// Graph Node Definitions (Discriminated Union)
// =============================================================================

export interface GraphNodeBase {
  id: string;
  name: string;
  description?: string | null;
  reads: string[];
  writes: string[];
}

export interface LLMGraphNode extends GraphNodeBase {
  kind: "llm";
  config: LLMNodeConfig;
}

export interface ToolGraphNode extends GraphNodeBase {
  kind: "tool";
  config: ToolNodeConfig;
}

export interface TransformGraphNode extends GraphNodeBase {
  kind: "transform";
  config: TransformNodeConfig;
}

export interface ComponentGraphNode extends GraphNodeBase {
  kind: "component";
  config: ComponentNodeConfig;
}

export type GraphNodeConfig =
  | LLMGraphNode
  | ToolGraphNode
  | TransformGraphNode
  | ComponentGraphNode;

// =============================================================================
// Graph Edge Definition
// =============================================================================

export interface EdgePredicate {
  state_path: string;
  operator: PredicateOperator;
  value?: unknown;
}

export interface GraphEdgeConfig {
  from_node: string; // Node ID, "START" is NOT stored in v3 edges
  to_node: string; // Node ID or "END"
  when?: BuiltinEdgeCondition | EdgePredicate | null;
  priority?: number;
  label?: string | null;
}

// =============================================================================
// Graph IR (Executable Body)
// =============================================================================

export interface GraphIR {
  nodes: GraphNodeConfig[];
  edges: GraphEdgeConfig[];
  entrypoints: string[];
}

// =============================================================================
// Prompt Configuration Types
// =============================================================================

export interface IdentityConfig {
  name?: string;
  description?: string;
  persona?: string | null;
}

export interface BrandingConfig {
  mask_provider?: boolean;
  mask_model?: boolean;
  branded_name?: string;
  forbidden_reveals?: string[];
}

export interface SecurityConfig {
  injection_defense?: boolean;
  refuse_prompt_reveal?: boolean;
  refuse_instruction_override?: boolean;
  confidential_sections?: string[];
}

export interface SafetyConfig {
  content_safety?: boolean;
  refuse_illegal?: boolean;
  refuse_harmful?: boolean;
  refuse_explicit?: boolean;
  refuse_violence?: boolean;
  refuse_hate?: boolean;
  refuse_self_harm?: boolean;
}

export interface FormattingConfig {
  use_markdown?: boolean;
  code_blocks?: boolean;
  language_identifiers?: boolean;
  custom_blocks?: string[];
}

export interface ContextConfig {
  include_date?: boolean;
  include_time?: boolean;
  date_format?: string;
  custom_context?: string | null;
}

export interface OverridesConfig {
  meta_instruction?: string | null;
  persona_instruction?: string | null;
  tool_instruction?: string | null;
  format_instruction?: string | null;
}

export interface PromptConfig {
  version?: string;
  identity?: IdentityConfig;
  branding?: BrandingConfig;
  security?: SecurityConfig;
  safety?: SafetyConfig;
  formatting?: FormattingConfig;
  context?: ContextConfig;
  custom_instructions?: string | null;
  overrides?: OverridesConfig;
}

// =============================================================================
// Complete Graph Configuration (v3 Canonical)
// =============================================================================

export interface GraphConfig {
  schema_version: "3.0";
  key: string;
  revision: number;
  graph: GraphIR;
  state: GraphStateConfig;
  deps?: GraphDeps | null;
  limits: GraphExecutionLimits;
  prompt_config?: PromptConfig | null;
  metadata?: GraphMetadata | null;
  /** UI-only metadata ignored by the compiler (e.g., node positions). */
  ui?: Record<string, unknown> | null;
}

// =============================================================================
// Helper Functions
// =============================================================================

/**
 * Create a default empty GraphConfig (v3 canonical).
 */
export function createEmptyGraphConfig(
  key: string = "custom_graph",
): GraphConfig {
  return {
    schema_version: "3.0",
    key,
    revision: 1,
    graph: {
      nodes: [],
      edges: [],
      entrypoints: [],
    },
    state: {},
    limits: { max_time_s: 300, max_steps: 128, max_concurrency: 10 },
  };
}

/**
 * Create a default LLM node.
 */
export function createDefaultLLMNode(
  id: string,
  name: string = "LLM Node",
): LLMGraphNode {
  return {
    id,
    name,
    kind: "llm",
    reads: ["messages"],
    writes: ["messages", "response"],
    config: {
      prompt_template: "",
      output_key: "response",
      tools_enabled: true,
      max_iterations: 10,
    },
  };
}

/**
 * Create a default Tool node.
 */
export function createDefaultToolNode(
  id: string,
  name: string = "Tool Node",
): ToolGraphNode {
  return {
    id,
    name,
    kind: "tool",
    reads: ["messages"],
    writes: ["tool_results"],
    config: {
      execute_all: true,
      output_key: "tool_results",
      timeout_seconds: 60,
    },
  };
}

/**
 * Create a default Transform node.
 */
export function createDefaultTransformNode(
  id: string,
  name: string = "Transform Node",
): TransformGraphNode {
  return {
    id,
    name,
    kind: "transform",
    reads: [],
    writes: ["output"],
    config: {
      template: "",
      output_key: "output",
      input_keys: [],
    },
  };
}

/**
 * Create a default Component node.
 */
export function createDefaultComponentNode(
  id: string,
  name: string = "Component Node",
): ComponentGraphNode {
  return {
    id,
    name,
    kind: "component",
    reads: [],
    writes: [],
    config: {
      component_ref: { key: "", version: "*" },
      config_overrides: {},
    },
  };
}

/**
 * Validate a GraphConfig and return errors.
 */
export function validateGraphConfig(config: GraphConfig): string[] {
  const errors: string[] = [];

  if (!config.graph) {
    errors.push("Missing 'graph' field");
    return errors;
  }

  const nodeIds = new Set(config.graph.nodes.map((n) => n.id));

  // Check entrypoints exist
  for (const ep of config.graph.entrypoints) {
    if (!nodeIds.has(ep)) {
      errors.push(`Entrypoint '${ep}' not found in nodes`);
    }
  }

  // Check edge references are valid
  const validRefs = new Set([...nodeIds, "END"]);
  for (const edge of config.graph.edges) {
    if (!nodeIds.has(edge.from_node)) {
      errors.push(`Edge from_node '${edge.from_node}' not found`);
    }
    if (!validRefs.has(edge.to_node)) {
      errors.push(`Edge to_node '${edge.to_node}' not found`);
    }
  }

  // Check each node has a valid kind
  const validKinds = new Set<string>(["llm", "tool", "transform", "component"]);
  for (const node of config.graph.nodes) {
    if (!validKinds.has(node.kind)) {
      errors.push(`Node '${node.id}' has invalid kind '${node.kind}'`);
    }
  }

  return errors;
}

/**
 * Get node kind display info.
 */
export function getNodeTypeInfo(kind: GraphNodeKind): {
  label: string;
  description: string;
  color: string;
  icon: string;
} {
  const info: Record<
    GraphNodeKind,
    { label: string; description: string; color: string; icon: string }
  > = {
    llm: {
      label: "LLM",
      description: "AI reasoning and generation",
      color: "#8b5cf6", // violet
      icon: "sparkles",
    },
    tool: {
      label: "Tool",
      description: "Execute external tools",
      color: "#3b82f6", // blue
      icon: "wrench",
    },
    transform: {
      label: "Transform",
      description: "Data transformation",
      color: "#6366f1", // indigo
      icon: "arrows-exchange",
    },
    component: {
      label: "Component",
      description: "Reusable component reference",
      color: "#059669", // emerald-600
      icon: "puzzle",
    },
  };

  return info[kind];
}

File: service/app/schemas/graph_config.py
"""
Graph Configuration Schema.

Canonical IR for executable graph definitions.
Enforces strict field boundaries and explicit entrypoints.
"""

from __future__ import annotations

from enum import StrEnum
from typing import Annotated, Any, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.schemas.prompt_config import PromptConfig


class GraphNodeKind(StrEnum):
    """Supported executable node kinds."""

    LLM = "llm"
    TOOL = "tool"
    TRANSFORM = "transform"
    COMPONENT = "component"


class StateFieldType(StrEnum):
    """Supported state field types."""

    STRING = "string"
    INT = "int"
    FLOAT = "float"
    BOOL = "bool"
    LIST = "list"
    DICT = "dict"
    ANY = "any"


class StateReducerType(StrEnum):
    """Supported state reducers."""

    REPLACE = "replace"
    ADD_MESSAGES = "add_messages"


class BuiltinEdgeCondition(StrEnum):
    """Built-in edge conditions."""

    HAS_TOOL_CALLS = "has_tool_calls"
    NO_TOOL_CALLS = "no_tool_calls"


class PredicateOperator(StrEnum):
    """Supported custom predicate operators."""

    EQUALS = "eq"
    NOT_EQUALS = "neq"
    TRUTHY = "truthy"
    FALSY = "falsy"


class StateFieldSchema(BaseModel):
    """State field schema definition."""

    model_config = ConfigDict(extra="forbid")

    type: StateFieldType = Field(description="State value type")
    description: str | None = Field(default=None, description="Human-readable description")
    default: Any = Field(default=None, description="Default value")


class GraphStateConfig(BaseModel):
    """State contract for the graph."""

    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    state_schema: dict[str, StateFieldSchema] = Field(
        default_factory=dict,
        alias="schema",
        description="State field schema by path/key",
    )
    reducers: dict[str, StateReducerType] = Field(
        default_factory=dict,
        description="Reducer configuration by state path/key",
    )


class GraphExecutionLimits(BaseModel):
    """Execution safeguards for loops and runaway workflows."""

    model_config = ConfigDict(extra="forbid")

    max_time_s: int = Field(default=300, ge=1, le=3600)
    max_steps: int = Field(default=128, ge=1, le=100000)
    max_concurrency: int = Field(default=10, ge=1, le=256)


class ModelDependencyRef(BaseModel):
    """Model dependency declaration."""

    model_config = ConfigDict(extra="forbid")

    key: str = Field(min_length=1)
    provider: str | None = None
    version: str | None = None


class PromptDependencyRef(BaseModel):
    """Prompt dependency declaration."""

    model_config = ConfigDict(extra="forbid")

    key: str = Field(min_length=1)
    version: str | None = None


class ComponentDependencyRef(BaseModel):
    """Component dependency declaration."""

    model_config = ConfigDict(extra="forbid")

    key: str = Field(min_length=1)
    version: str = Field(default="*")


class GraphDeps(BaseModel):
    """External references used by this graph definition."""

    model_config = ConfigDict(extra="forbid")

    models: list[ModelDependencyRef] = Field(default_factory=list)
    tools: list[str] = Field(default_factory=list)
    prompts: list[PromptDependencyRef] = Field(default_factory=list)
    components: list[ComponentDependencyRef] = Field(default_factory=list)


class GraphMetadata(BaseModel):
    """Display metadata that does not affect execution semantics."""

    model_config = ConfigDict(extra="forbid")

    display_name: str | None = None
    description: str | None = None
    tags: list[str] = Field(default_factory=list)
    agent_version: str | None = None


class LLMNodeConfig(BaseModel):
    """LLM node runtime configuration."""

    model_config = ConfigDict(extra="forbid")

    prompt_template: str = ""
    output_key: str = "response"
    model_override: str | None = None
    temperature_override: float | None = Field(default=None, ge=0.0, le=2.0)
    max_tokens: int | None = None
    tools_enabled: bool = True
    tool_filter: list[str] | None = None
    max_iterations: int = Field(default=10, ge=1)
    message_key: str | None = None


class ToolNodeConfig(BaseModel):
    """Tool node runtime configuration."""

    model_config = ConfigDict(extra="forbid")

    execute_all: bool = True
    tool_filter: list[str] | None = None
    output_key: str = "tool_results"
    timeout_seconds: int = Field(default=60, ge=1, le=600)


class TransformNodeConfig(BaseModel):
    """Transform node runtime configuration."""

    model_config = ConfigDict(extra="forbid")

    template: str
    output_key: str
    input_keys: list[str] = Field(default_factory=list)


class ComponentReference(BaseModel):
    """Reference to a registered executable component."""

    model_config = ConfigDict(extra="forbid")

    key: str = Field(min_length=1)
    version: str = Field(default="*")


class ComponentNodeConfig(BaseModel):
    """Component node runtime configuration."""

    model_config = ConfigDict(extra="forbid")

    component_ref: ComponentReference
    config_overrides: dict[str, Any] = Field(default_factory=dict)


class GraphNodeBase(BaseModel):
    """Common node fields."""

    model_config = ConfigDict(extra="forbid")

    id: str = Field(min_length=1)
    name: str = Field(min_length=1)
    description: str | None = None
    reads: list[str] = Field(default_factory=list)
    writes: list[str] = Field(default_factory=list)


class LLMGraphNode(GraphNodeBase):
    """LLM node."""

    kind: Literal[GraphNodeKind.LLM]
    config: LLMNodeConfig


class ToolGraphNode(GraphNodeBase):
    """Tool node."""

    kind: Literal[GraphNodeKind.TOOL]
    config: ToolNodeConfig


class TransformGraphNode(GraphNodeBase):
    """Transform node."""

    kind: Literal[GraphNodeKind.TRANSFORM]
    config: TransformNodeConfig


class ComponentGraphNode(GraphNodeBase):
    """Component node."""

    kind: Literal[GraphNodeKind.COMPONENT]
    config: ComponentNodeConfig


GraphNodeConfig = Annotated[
    LLMGraphNode | ToolGraphNode | TransformGraphNode | ComponentGraphNode,
    Field(discriminator="kind"),
]


class EdgePredicate(BaseModel):
    """Custom edge predicate on state."""

    model_config = ConfigDict(extra="forbid")

    state_path: str = Field(min_length=1)
    operator: PredicateOperator
    value: Any = None


class GraphEdgeConfig(BaseModel):
    """Directed edge in the graph."""

    model_config = ConfigDict(extra="forbid")

    from_node: str = Field(min_length=1, description="Node ID or START")
    to_node: str = Field(min_length=1, description="Node ID or END")
    when: BuiltinEdgeCondition | EdgePredicate | None = Field(
        default=None,
        description="Conditional predicate or built-in routing condition",
    )
    priority: int = 0
    label: str | None = None


class GraphIR(BaseModel):
    """Executable graph body."""

    model_config = ConfigDict(extra="forbid")

    nodes: list[GraphNodeConfig]
    edges: list[GraphEdgeConfig]
    entrypoints: list[str] = Field(min_length=1)

    @field_validator("entrypoints")
    @classmethod
    def _validate_entrypoints_unique(cls, entrypoints: list[str]) -> list[str]:
        if len(set(entrypoints)) != len(entrypoints):
            raise ValueError("graph.entrypoints must be unique")
        return entrypoints


class GraphConfig(BaseModel):
    """Canonical graph configuration."""

    model_config = ConfigDict(extra="forbid")

    schema_version: Literal["3.0"] = "3.0"
    key: str = Field(min_length=1)
    revision: int = Field(default=1, ge=1)
    graph: GraphIR
    state: GraphStateConfig = Field(default_factory=GraphStateConfig)
    deps: GraphDeps | None = None
    limits: GraphExecutionLimits = Field(default_factory=GraphExecutionLimits)

    # Kept in phase-1 for prompt compatibility with current runtime.
    prompt_config: PromptConfig | None = None

    metadata: GraphMetadata | None = None
    ui: dict[str, Any] | None = Field(
        default=None,
        description="UI-only metadata ignored by compiler",
    )


def is_graph_config(config: dict[str, Any]) -> bool:
    """Quick version check for raw graph config payloads."""

    return str(config.get("schema_version", "")).startswith("3.")


def parse_graph_config(config: dict[str, Any]) -> GraphConfig:
    """Parse and validate raw payload as GraphConfig."""

    return GraphConfig.model_validate(config)


__all__ = [
    "BuiltinEdgeCondition",
    "ComponentDependencyRef",
    "ComponentNodeConfig",
    "ComponentReference",
    "EdgePredicate",
    "GraphConfig",
    "GraphDeps",
    "GraphEdgeConfig",
    "GraphExecutionLimits",
    "GraphIR",
    "GraphMetadata",
    "GraphNodeConfig",
    "GraphNodeKind",
    "GraphStateConfig",
    "LLMNodeConfig",
    "ModelDependencyRef",
    "PredicateOperator",
    "PromptDependencyRef",
    "StateFieldSchema",
    "StateFieldType",
    "StateReducerType",
    "ToolNodeConfig",
    "TransformNodeConfig",
    "is_graph_config",
    "parse_graph_config",
]
